File: languages/rust/reverse-complement.rs
// reverse-complement: generate a DNA sequence, reverse it in place while complementing
// each base (A<->T, C<->G), then reduce it to a polynomial string hash. The reverse uses a
// hand-written two-pointer loop (NOT a stdlib bulk reverse) and the hash a per-character
// loop (NOT a builtin), so this measures the language's own per-character processing -
// consistent with the suite's no-stdlib-shortcut rule. Everything is integer-deterministic.
use std::env;

const P: i64 = 1000000007;
const IM: i64 = 139968;
const IA: i64 = 3877;
const IC: i64 = 29573;

fn comp(c: u8) -> u8 {
    // A<->T, C<->G; only A/C/G/T occur
    if c == b'A' {
        b'T'
    } else if c == b'C' {
        b'G'
    } else if c == b'G' {
        b'C'
    } else {
        b'A'
    }
}

fn run(l: usize) -> i64 {
    let mut s: Vec<u8> = vec![0u8; l];
    let mut seed: i64 = 42;
    for i in 0..l {
        seed = (seed * IA + IC) % IM;
        s[i] = if seed < 42000 {
            b'A'
        } else if seed < 70000 {
            b'C'
        } else if seed < 98000 {
            b'G'
        } else {
            b'T'
        };
    }

    // two-pointer reverse-and-complement, in place
    let mut i: usize = 0;
    if l > 0 {
        let mut j: usize = l - 1;
        while i < j {
            let a = comp(s[i]);
            s[i] = comp(s[j]);
            s[j] = a;
            i += 1;
            j -= 1;
        }
        if i == j {
            s[i] = comp(s[i]); // middle char when L is odd
        }
    }

    // polynomial string hash over the ASCII byte values (A=65, C=67, G=71, T=84)
    let mut h: i64 = 0;
    for k in 0..l {
        h = (h * 31 + s[k] as i64) % P;
    }
    h
}

fn main() {
    let l: usize = env::args().nth(1).and_then(|s| s.parse().ok()).unwrap_or(100000);
    println!("{}", run(l));
    println!("reverse-complement({})", l);
}

File: languages/ruby/reverse-complement.rb
# reverse-complement: generate a DNA sequence, reverse it in place while complementing
# each base (A<->T, C<->G), then reduce it to a polynomial string hash. The reverse uses a
# hand-written two-pointer loop (NOT a stdlib bulk reverse) and the hash a per-character
# loop (NOT a builtin), so this measures the language's own per-character processing -
# consistent with the suite's no-stdlib-shortcut rule. Everything is integer-deterministic.

P = 1000000007
IM = 139968
IA = 3877
IC = 29573

def comp(c) # A<->T, C<->G; only A/C/G/T occur
  return 84 if c == 65 # 'A' -> 'T'
  return 71 if c == 67 # 'C' -> 'G'
  return 67 if c == 71 # 'G' -> 'C'
  65                   # 'T' -> 'A'
end

def reverse_complement(l)
  s = Array.new(l, 0)
  seed = 42
  i = 0
  while i < l
    seed = (seed * IA + IC) % IM
    s[i] = if seed < 42000
             65 # 'A'
           elsif seed < 70000
             67 # 'C'
           elsif seed < 98000
             71 # 'G'
           else
             84 # 'T'
           end
    i += 1
  end

  i = 0
  j = l - 1
  while i < j # two-pointer reverse-and-complement, in place
    a = comp(s[i])
    s[i] = comp(s[j])
    s[j] = a
    i += 1
    j -= 1
  end
  s[i] = comp(s[i]) if i == j # middle char when L is odd

  h = 0
  k = 0
  while k < l
    h = (h * 31 + s[k]) % P
    k += 1
  end
  h
end

l = ARGV[0] ? ARGV[0].to_i : 400000
puts reverse_complement(l)
puts "reverse-complement(#{l})"
